### PR TITLE
Compat: Test textureBindingViewDimension validation

### PR DESCRIPTION
**Pull request by greggman**
Tests that if textureBindingViewDimension is incompatible with a texture's dimension in createTexture a validation error is generated

Tests that if a texture view's dimension does not match the texture's textureBindingViewDimension when passed to createBindGroup a validation error is generated.

<hr>

**Requirements for PR author:**

- [X] All missing test coverage is tracked with "TODO" or `.unimplemented()`.
- [X] New helpers are `/** documented */` and new helper files are found in `helper_index.txt`.
- [X] Test behaves as expected in a WebGPU implementation. (If not passing, explain above.)

**Requirements for [reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md):**

- [ ] Tests are properly located in the test tree.
- [ ] [Test descriptions](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) allow a reader to "read only the test plans and evaluate coverage completeness", and accurately reflect the test code.
- [ ] Tests provide complete coverage (including validation control cases). **Missing coverage MUST be covered by TODOs.**
- [ ] Helpers and types promote readability and maintainability.

When landing this PR, be sure to make any necessary issue status updates.

### CODE REVIEW

**Comment by greggman** [comment]
These pass in Chrome with the proper validating compiled in.

File: src/webgpu/compat/api/validation/createBindGroup.spec.ts [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+import { CompatibilityTest } from '../../compatibility_test.js';
+
+export const g = makeTestGroup(CompatibilityTest);
+
+function isTextureBindingViewDimensionCompatibleWithDimension(
+  dimension?: GPUTextureDimension,
+  textureBindingViewDimension?: GPUTextureViewDimension
+) {
+  dimension = dimension || '2d';
+  textureBindingViewDimension = textureBindingViewDimension || '2d';
```
Comment (kainino0x) [inline src/webgpu/compat/api/validation/createBindGroup.spec.ts:20]:
```suggestion
  dimension: GPUTextureDimension = '2d',
  textureBindingViewDimension: GPUTextureViewDimension = '2d'
) {
```

File: src/webgpu/compat/api/validation/createBindGroup.spec.ts [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+      )
+  )
+  .fn(t => {
+    const { dimension, textureBindingViewDimension, viewDimension, depthOrArrayLayers } = t.params;
+
+    const texture = t.device.createTexture({
+      size: [1, 1, depthOrArrayLayers],
+      format: 'rgba8unorm',
+      usage: GPUTextureUsage.TEXTURE_BINDING,
+      ...(dimension && { dimension }),
```
Comment (kainino0x) [inline src/webgpu/compat/api/validation/createBindGroup.spec.ts:139]: weird, but I'm going to have to file this one in my book of JS tricks.

We usually don't bother making sure unused keys won't be set to `undefined` - webidl `dictionary` guarantees that `{}` and `{x: undefined}` are the same. IMO probably not necessary here, but OK.

File: src/webgpu/compat/api/validation/createBindGroup.spec.ts [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+    case '2d':
+    case undefined:
+      return depthOrArrayLayers > 1 ? '2d-array' : '2d';
+      break;
+    case '3d':
+      return '3d';
+  }
+}
+
+g.test('dimension_matches_textureBindingViewDimension')
```
Comment (kainino0x) [inline src/webgpu/compat/api/validation/createBindGroup.spec.ts:104]: iiuc
```suggestion
g.test('viewDimension_matches_textureBindingViewDimension')
```

File: src/webgpu/compat/api/validation/createBindGroup.spec.ts [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+        },
+      ],
+    });
+
+    const shouldError = effectiveTextureBindingViewDimension !== effectiveViewDimension;
+
+    t.expectValidationError(() => {
+      t.device.createBindGroup({
+        layout,
+        entries: [{ binding: 0, resource: texture.createView({ dimension: viewDimension }) }],
```
Comment (kainino0x) [inline src/webgpu/compat/api/validation/createBindGroup.spec.ts:173]: It's important especially for this test to check whether the error comes from createView or createBindGroup. I'm assuming createBindGroup, so the createView should be moved out of the expectValidationError.

File: src/webgpu/compat/api/validation/createBindGroup.spec.ts [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+      texture.depthOrArrayLayers,
+      viewDimension
+    );
+
+    const layout = t.device.createBindGroupLayout({
+      entries: [
+        {
+          binding: 0,
+          visibility: GPUShaderStage.COMPUTE,
+          texture: {
```
Comment (kainino0x) [inline src/webgpu/compat/api/validation/createBindGroup.spec.ts:161]: Would be good to verify this isn't enforced for storage textures

File: src/webgpu/util/texture/base.ts [hunk 6]
```diff
@@ -157,22 +157,52 @@ export function viewDimensionsForTextureDimension(textureDimension: GPUTextureDi
...
     case '3d':
       return '3d';
     default:
       unreachable();
   }
 }
 
+/** Returns the effective view dimension for a given texture */
+export function effectiveViewDimensionForTexture(
+  texture: Readonly<GPUTexture>,
```
Comment (kainino0x) [inline src/webgpu/util/texture/base.ts:186]: nit: GPUTexture should be inherently readonly, I think
```suggestion
  texture: GPUTexture,
```

**Review by kainino0x — APPROVED** [review]
Sorry for the delay.

File: src/webgpu/compat/api/validation/createBindGroup.spec.ts (hunk 5 above)
Comment (greggman) [reply src/webgpu/compat/api/validation/createBindGroup.spec.ts:161]: The storage textures should (already?) be tested by core. The compat specific tests only test that things not allowed are validated out in compat. For things that still work they should be tested already (or in any case belong in the storage texture tests... right?)

**Pull request merged**